CCO-21: Support AWS Security Token Service (4.8 GA)

jeana-redhat opened the pull request:
For https://issues.redhat.com/browse/OSDOCS-2196
Preview: https://deploy-preview-31136--osdocs.netlify.app/openshift-enterprise/latest/authentication/managing_cloud_provider_credentials/cco-mode-sts.html#sts-mode-installing

jeana-redhat commented:
@joelddiaz this is ready for a look, when you get a chance

jeana-redhat commented on modules/cco-ccoctl-creating-at-once.adoc:
```diff
@@ -0,0 +1,60 @@
++
+[source,terminal,subs="+quotes"]
+----
+$ ccoctl aws create-all --name-prefix=__<name_prefix>__ --region=__<aws_region>__ --credentials-requests-dir=./credrequests
```
Digging around in Jira this afternoon I see s/name-prefix/name, will update throughout

jeana-redhat replied:
Related Q: what does "tag and account any cloud resources" mean? Specifically the "account" part.

joelddiaz replied:
Perhaps it would be better worded as "tag for tracking". Basically, the tool names things using the `--name` flag, and tags them using that field as well.

joelddiaz commented on modules/cco-ccoctl-creating-at-once.adoc:
```diff
@@ -0,0 +1,60 @@
+----
+
+. Process every `CredentialsRequest` object in the `./credrequests` directory with `ccoctl`:
+// My assumption is that it does all `CredentialsRequest` objects rather than "each" as in the other procedure.
```
Yes, it reads each file in that directory. And processes them one at a time (generating the IAM Roles and the OpenShift Secret files).

joelddiaz commented on modules/cco-ccoctl-creating-at-once.adoc:
```diff
@@ -0,0 +1,60 @@
+
+.Verification
+
+You can verify that the IAM roles are created by checking the secret files in the `_<path_to_ccoctl_output_dir>_/manifests` directory.
```
Actually, these files are the OpenShift Secrets that will be passed to the new cluster. The IAM Roles themselves would be visible by querying AWS directly (with something like `aws iam list-roles | jq -r .Roles[].RoleName`).

joelddiaz commented on modules/cco-ccoctl-creating-individually.adoc:
```diff
@@ -0,0 +1,122 @@
+[id="cco-ccoctl-creating-individually_{context}"]
+= Creating AWS resources individually
+
+// Would be nice to have a little explanation why a user would want to go this route vs. the other option.
```
The details in this section is basically pulling back the curtain and showing how this all really works under the covers. I think the main use case here is that the individual/company wants to delegate responsibility for creating the various objects across different company departments.

joelddiaz commented on modules/cco-ccoctl-creating-individually.adoc:
```diff
@@ -0,0 +1,122 @@
++
+Where `02-openid-configuration` is a discovery document and `03-keys.json` is a JSON web key set file.
+// Lin's output has openid-configuration (not 02-openid-configuration) and keys.json (not 03-keys.json) - does something in this output need correction?
+// I believe based on AWS docs that the string `301721915996` in the test plan is the AWS account ID, but would like confirmation on that.
```
Yes, that is the length of an AWS account id.

joelddiaz commented on modules/cco-ccoctl-creating-individually.adoc:
```diff
@@ -0,0 +1,122 @@
+// I believe based on AWS docs that the string `301721915996` in the test plan is the AWS account ID, but would like confirmation on that.
++
+This command also creates a YAML configuration file in `/_<path_to_ccoctl_output_dir>_/manifests/cluster-authentication-02-config.yaml`.
+// Does the user need to know more about this?
```
If you want to provide details, this will set the IssuerURL field for the ServiceAccount tokens that the cluster will be generating. Which is required to be properly set so that the AWS IAM Identity Provider created above will recognize/trust the tokens.

joelddiaz commented on modules/cco-ccoctl-creating-individually.adoc:
```diff
@@ -0,0 +1,122 @@
+----
+$ oc adm release extract --credentials-requests --cloud=aws --to=./credrequests quay.io/__<path_to>__/openshift-release:__<version>__
+----
+// Can `--to=./credrequests` be replaced with `--to=<path_to_directory_with_list_of_credentials_requests>/credrequests`?
```
Yes. basically just a path to a directory that the `oc` command will dump its output into. So it can be anything that is a valid path.

joelddiaz commented on modules/cco-ccoctl-creating-individually.adoc:
```diff
@@ -0,0 +1,122 @@
+// Can `--to=./credrequests` be replaced with `--to=<path_to_directory_with_list_of_credentials_requests>/credrequests`?
+
+.. Process each `CredentialsRequest` object in the `./credrequests` directory with `ccoctl`:
+// Does the user have to repeat this for each? OR is it getting all at once from the directory?
```
Process "all CredentialsRequest objects in the './credrequests' directory..."

joelddiaz commented on modules/cco-ccoctl-creating-individually.adoc:
```diff
@@ -0,0 +1,122 @@
++
+[NOTE]
+====
+For AWS GovCloud, you must also specify your region with the `--region` parameter.
```
```suggestion
For any AWS environment that uses alternative IAM API endpoints (eg. GovCloud), you must also specify your region with the `--region` parameter.
```

joelddiaz commented on modules/cco-ccoctl-creating-individually.adoc:
```diff
@@ -0,0 +1,122 @@
+
+.Verification
+
+You can verify that the IAM roles are created by checking the secret files in the `_<path_to_ccoctl_output_dir>_/manifests` directory.
```
Same comment as before. To get the llist of IAM Roles you need to query AWS directly. This is just the Secrets that will be placed into the cluster.

joelddiaz commented on modules/sts-mode-installing-manual-run-installer.adoc:
```diff
@@ -5,7 +5,40 @@
+----
+
+. Copy the generated manifests to the installation directory:
+// Confused by this step from the enablement slide. Isn't the previous step run from the right directory already (if we're in the installation directory that contains the config YAML)?
```
What needs to happen is that we save the manifest files created by `ccoctl` to the side, then allow the installer to do its thing and create its own separate manifests. Then we can finally merge the contents of the two directories by adding only the `ccoctl` manifests files next to the installer generated manifests files.

jeana-redhat replied:
aaaah ok, so `Copy the generated manifests to the installation directory` is `Copy the manifests that ccoctl generated to the manifests directory that the installation program created`

joelddiaz reviewed (commented):
looks good. couple of clarifications/comments.

jeana-redhat commented on modules/cco-ccoctl-configuring.adoc:
```diff
@@ -0,0 +1,79 @@
+.Prerequisites
+
+* Obtain the {product-title} release image.
+// Pulling the ccoctl tool from the release image is supported in Linux only, so can this be done outside Linux?
```
Probably best to just cover the Linux case entirely then, to avoid extra if/then complications.

jeana-redhat commented on modules/cco-ccoctl-configuring.adoc:
```diff
@@ -0,0 +1,79 @@
+----
+
+. Extract `ccoctl` from the release image:
+// Is that what we're doing?
```
ok, attempted to clarify a bit

jeana-redhat commented on modules/cco-ccoctl-configuring.adoc:
```diff
@@ -0,0 +1,79 @@
+
+Use "ccoctl aws [command] --help" for more information about a command.
+----
+// Is there a reason that the help file for this doesn't list --target-dir and --generate-files-only under "Flags"? Does it matter?
```
Ah - that makes sense. I like how oc has these consistent recursive help trees :) Will update flags throughout.

jeana-redhat commented on modules/cco-ccoctl-creating-at-once.adoc:
```diff
@@ -0,0 +1,60 @@
+[id="cco-ccoctl-creating-at-once_{context}"]
+= Creating AWS resources with a single command
+
+// Would be nice to have a little explanation why a user would want to go this route vs. the other option.
```
Thanks, that helps a lot! So I wonder now if we actually need to have the part about `--dry-run` in the note on the all-at-once procedure?

joelddiaz replied:
I can't imagine why someone would do --dry-run with the create-all command... And it doesn't appear that the create-all will recognize the --dry-run parameter.

jeana-redhat commented:
Updated with feedback, and added context about when a user might prefer each option. I think the only remaining quesion I had is about the discrepancy between `openid-configuration` and `keys.json` in Lin's test output vs. `02-openid-configuration` and `03-keys.json` [here](https://github.com/openshift/openshift-docs/pull/31136/files#diff-8292a813a752018e3e3cc23f20dfea97c7892161b18b7faba24bc7895071902eR73).
@joelddiaz PTAL at the revised docs and that particular question when you get a chance :pray:

joelddiaz commented on modules/sts-mode-installing-manual-run-installer.adoc:
```diff
@@ -5,7 +5,39 @@
+$ cp /<path_to_ccoctl_output_dir>/manifests/* ./manifests/
+----
+
+. Copy the manifests that the `ccoctl` generated in the `tls` directory to the installation directory:
```
```suggestion
. Copy the private key that the `ccoctl` generated in the `tls` directory to the installation directory:
```

joelddiaz reviewed (commented):
one tiny nit. otherwise looks great

jeana-redhat commented:
Hi @lwan-wanglin - this PR is to cover the installation process for a cluster using STS in 4.8. Joel is happy with it, so please take a look when you get a chance.

lwan-wanglin commented on modules/cco-ccoctl-creating-individually.adoc:
```diff
@@ -0,0 +1,124 @@
++
+[source,terminal]
+----
+$ ccoctl create key-pair
```
This command should be `$ ccoctl aws create-key-pair`

lwan-wanglin commented on modules/cco-ccoctl-creating-individually.adoc:
```diff
@@ -0,0 +1,124 @@
++
+** `_<name>_` is the name used to tag any cloud resources that are created for tracking.
+** `_<aws-region>_` is the AWS region in which cloud resources will be created.
+** `_<path_to_ccoctl_output_dir>_` is the path to the public key file that the `ccoctl create key-pair` command generated.
```
Should be:
_<path_to_ccoctl_output_dir>_` is the path to the public key file that the **ccoctl aws create-key-pair** command generated

lwan-wanglin commented on modules/cco-ccoctl-creating-individually.adoc:
```diff
@@ -0,0 +1,124 @@
++
+[source,terminal,subs="+quotes"]
+----
+$ ccoctl create identity-provider --name=__<name>__ --region=__<aws_region>__ --public-key=__<path_to_ccoctl_output_dir>__/serviceaccount-signer.public
```
This command should be `$ ccoctl aws create-identity-provider --name=__<name>__ --region=__<aws_region>__ --public-key-file=__<path_to_ccoctl_output_dir>__/serviceaccount-signer.public`

lwan-wanglin commented on modules/cco-ccoctl-creating-at-once.adoc:
```diff
@@ -0,0 +1,63 @@
++
+[source,terminal,subs="+quotes"]
+----
+$ oc adm release extract --credentials-requests --cloud=aws --to=./credrequests quay.io/__<path_to>__/openshift-release:__<version>__
```
it would be better to keep the content consistent with line 83 in the doc cco-ccoctl-creating-individually.adoc

`oc adm release extract --credentials-requests --cloud=aws --to=<path_to_directory_with_list_of_credentials_requests>/credrequests quay.io/<path_to>/openshift-release:<version>`

lwan-wanglin commented on modules/cco-ccoctl-creating-at-once.adoc:
```diff
@@ -0,0 +1,63 @@
++
+[source,terminal,subs="+quotes"]
+----
+$ ccoctl aws create-all --name=__<name>__ --region=__<aws_region>__ --credentials-requests-dir=./credrequests
```
same with the above comment, using "<path_to_directory_with_list_of_credentials_requests>/credrequest" instead "./credrequests"

jeana-redhat commented:
Thanks, @lwan-wanglin - refreshed push has your changes :)

bobfuru commented on authentication/managing_cloud_provider_credentials/cco-mode-sts.adoc:
```diff
@@ -60,17 +57,38 @@ data:
+include::modules/cco-ccoctl-configuring.adoc[leveloffset=+2]
+
+[id="sts-mode-create-aws-resources-ccoctl"]
+=== Creating AWS resources with the Cloud Credential Operator utility
```
Everything looks good on the front end but wondering why this part isn't a module? I'm sure it's for good reason but can't wrap my head around it. 😀

bobfuru commented on authentication/managing_cloud_provider_credentials/cco-mode-sts.adoc:
```diff
@@ -60,17 +57,38 @@ data:
 == Installing an {product-title} cluster configured for manual mode with STS
 
-To install a cluster that is configured to use the CCO in manual mode with STS in {product-title} version 4.7:
+To install a cluster that is configured to use the Cloud Credential Operator (CCO) in manual mode with STS:
```
Do the ordered steps in this super-task need periods?

bobfuru commented on modules/cco-ccoctl-configuring.adoc:
```diff
@@ -0,0 +1,73 @@
+
+[NOTE]
+====
+The `ccoctl` is a Linux binary, and must be run in a Linux environment.
```
Maybe this sounds better than "be run"?

```suggestion
The `ccoctl` is a Linux binary that must run in a Linux environment.
```

bobfuru commented on modules/cco-ccoctl-creating-at-once.adoc:
```diff
@@ -0,0 +1,63 @@
+[id="cco-ccoctl-creating-at-once_{context}"]
+= Creating AWS resources with a single command
+
+If you do not need to review the JSON files that the `ccoctl` creates before modifying AWS resources, and if the process the `ccoctl` uses to create AWS resources automatically meets the requirements of your organization, you can use the `ccoctl aws create-all` command to automate the creation of AWS resources.
```
Instead of this:

> that the `ccoctl` creates

Consider this:

> that the `ccoctl` tool creates

or this:

> that `ccoctl` creates

bobfuru replied:
Similarly, s/the process the `ccoctl` uses/the process that `ccoctl` uses/

bobfuru commented on modules/cco-ccoctl-creating-at-once.adoc:
```diff
@@ -0,0 +1,63 @@
+$ oc adm release extract --credentials-requests --cloud=aws --to=__<path_to_directory_with_list_of_credentials_requests>__/credrequests quay.io/__<path_to>__/openshift-release:__<version>__
+----
+
+. Process all `CredentialsRequest` objects in the `credrequests` directory with `ccoctl`:
```
To avoid ambiguity, consider:
s/with `ccoctl`:/by using `ccoctl`:/

jeana-redhat replied:
Flipped it around: 
Use the `ccoctl` tool to process all `CredentialsRequest` objects in the `credrequests` directory:

bobfuru commented on modules/cco-ccoctl-creating-at-once.adoc:
```diff
@@ -0,0 +1,63 @@
+
+.Verification
+
+You can verify that the {product-title} secrets are created by checking the secret files in the `_<path_to_ccoctl_output_dir>_/manifests` directory.
```
Not sure if the same rules apply for `Verification` as they do for `Procedure`, but usually we suggest that an intro sentence comes before the header. So after the heading, it's just a list of ordered steps. In this case, it doesn't make sense to have an intro sentence before `Verification` because it's a sub-part of the Procedure. `¯\_(ツ)_/¯`

One solution is to eliminate this intro sentence and in the step, write:

> List the secret files in the `<path_to_ccoctl_output_dir>/manifests` directory to verify that the {product-title} secrets were created.

jeana-redhat replied:
That's actually a good point, especially since the intro and the step are fairly similar. Thanks!

bobfuru commented on modules/cco-ccoctl-creating-individually.adoc:
```diff
@@ -0,0 +1,123 @@
+[id="cco-ccoctl-creating-individually_{context}"]
+= Creating AWS resources individually
+
+If you need to review the JSON files that the `ccoctl` creates before modifying AWS resources, or if the process the `ccoctl` uses to create AWS resources automatically does not meet the requirements of your organization, you can create the AWS resources individually. For example, this option might be useful for an organization that shares the responsibility for creating these resources among dif ...
```
Same comment as above re: `ccoctl`

bobfuru commented on modules/cco-ccoctl-creating-individually.adoc:
```diff
@@ -0,0 +1,123 @@
+
+.Procedure
+
+. Generate the public and private RSA key files are used to set up the OpenID Connect provider for the cluster:
```
"that are used"

bobfuru commented on modules/cco-ccoctl-creating-individually.adoc:
```diff
@@ -0,0 +1,123 @@
++
+Where `serviceaccount-signer.private` and `serviceaccount-signer.public` are the generated key files.
++
+This command also creates a private key that the cluster requires during installation in `/_<path_to_ccoctl_output_dir>_/tls/bound-service-account-signing-key.key`.
```
Double-checking that `-key.key` is correct here.

bobfuru commented on modules/cco-ccoctl-creating-individually.adoc:
```diff
@@ -0,0 +1,123 @@
+$ oc adm release extract --credentials-requests --cloud=aws --to=__<path_to_directory_with_list_of_credentials_requests>__/credrequests quay.io/__<path_to>__/openshift-release:__<version>__
+----
+
+.. Process all `CredentialsRequest` objects in the `./credrequests` directory with `ccoctl`:
```
s/with/by using/

bobfuru commented on modules/cco-ccoctl-creating-individually.adoc:
```diff
@@ -0,0 +1,123 @@
+
+.Verification
+
+You can verify that the {product-title} secrets are created by checking the secret files in the `_<path_to_ccoctl_output_dir>_/manifests` directory.
```
Same as previous comment under Verification.

bobfuru commented on modules/sts-mode-installing-manual-run-installer.adoc:
```diff
@@ -5,7 +5,39 @@
+. Copy the private key that the `ccoctl` generated in the `tls` directory to the installation directory:
++
+----
+$ cp -a /<path_to_ccoctl_output_dir>/tls .
```
In another part of the doc, you italicized the path variable so maybe it should be italicized here and in step 3?

jeana-redhat replied:
yes - and it didn't even have a source tag! oops

bobfuru reviewed (commented):
A few comments for your consideration but overall this looks great!

jeana-redhat commented:
/cherrypick enterprise-4.8

bobfuru commented on modules/cco-ccoctl-creating-individually.adoc:
```diff
@@ -0,0 +1,121 @@
+$ ccoctl aws create-identity-provider --name=__<name>__ --region=__<aws_region>__ --public-key-file=__<path_to_ccoctl_output_dir>__/serviceaccount-signer.public
+----
++
+Where:
```
I couldn't recall whether we were supposed to use upper or lower: `Where:` or `where:` but I'm now seeing ISG and [our own guidelines](https://github.com/openshift/openshift-docs/blob/master/contributing_to_docs/doc_guidelines.adoc#code-blocks-command-syntax-and-example-output) are using `where:`

jeana-redhat replied:
Thanks! I'll pull it into a subsequent PR :)

jeana-redhat replied:
in #32722 :+1:

jeana-redhat commented on modules/cco-ccoctl-creating-individually.adoc:
```diff
@@ -0,0 +1,121 @@
++
+[source,terminal]
+----
+$ ccoctl aws create key-pair
```
Updated in a later PR - current version is here: https://docs.openshift.com/container-platform/4.8/authentication/managing_cloud_provider_credentials/cco-mode-sts.html#cco-ccoctl-creating-individually_cco-mode-sts
Thanks for flagging this in any case!
